content/blog/2023-08-07-beginners-kernel-CTF-with-CVE-2017-5123.markdown

```diff
@@ -0,0 +1,380 @@
+(0) into the address in `r12`.
+
+(we write zero to the address placed at 32 / 0x20)
+
+Then, we replace the contents of `r12` with the value contained at address 24.
+Then, we move that value into the count register, and branch to the count
+register.
+
+(indirect branch to the address placed at 24 / 0x18)
+
+I wrote it this way so participants would have to understand what the shellcode
+was trying to do to be able to get any use out of it.  It expects two addresses
+to be placed immediately after it terminates and it's up to you to figure out
+what those addresses should be!
+
+In our case, everyone figured out pretty quickly that *24 should point at our
+very classic privesc:
+
+```c
+void get_root() {
+	if (commit_creds && prepare_kernel_cred)
+		commit_creds(prepare_kernel_cred(0));
+}
+```
+
+Addresses for those kernel symbols need to be obtained first, but we're experts
+at that now.  So we add in:
+
+```c
+	*(unsigned long *)24 = (unsigned long)get_root;
+```
+
+And that part's sorted.  How good is C?
+
+Noone guessed what address we were zeroing, though, and the answer is
+`have_canfork_callback`.  Without mending that, the kernel will keep attempting
+to execute from address 0, which we don't want.  We only need it to do that
+once!
+
+So we wrap up with
+
+```c
+	*(unsigned long *)32 = have_canfork_callback;
+```
+
+and our shellcode's ready to go!
+
+## Stage 6: it doesn't work
+
+We've had good progress so far - we needed a way to get the kernel to execute
+from address 0 and we found a way to do that, and we needed to `mmap` to 0 and
+we found a way to do that.  And yet, running the exploit doesn't work.  How
+come?
+
+```
+Unable to handle kernel paging request for instruction fetch
+Faulting instruction address: 0x00000000
+Oops: Kernel access of bad area, sig: 11 [#2]
+```
+
+The MMU has ended our fun.  KUEP is enabled (SMEP on x86, PXN on ARM) so the MMU
+is enforcing that the kernel can't execute from user addresses.  I gave everyone
+a bit of a trick question here - how can you get around this purely from the
+`qemu` command line?
+
+The way I did it wasn't to parse `nosmep` (and I'm not even sure that was
+implemented for powerpc in v4.13 anyway), it was to change from `-cpu POWER9` to
+`-cpu POWER8`.  Userspace execution prevention wasn't implemented in the MMU
+until POWER9, so reverting to an older processor was a cheeky way to get around
+that.
+
+## Stage 7: victory!
+
+Putting all of that together, we have a successful privilege escalation from
+attacking the kernel.
+
+```
+/ $ ./exploit 
+Overwriting mmap_min_addr...
+Overwriting dac_mmap_min_addr...
+Overwriting have_canfork_callback...
+Successfully acquired root shell!
+/ # whoami
+root
+```
+
+It's wild to think that even an exploit this simple would have been possible in
+the "real world" back in 2017.  Still, it made for a good introduction to kernel
+exploitation for me and my team and wasn't *too* contrived for the sake of
+simplicity.
```

Comment:
can we emphasise that this is why mitigations are valuable

Reply:
yeah ok